File: apps/android/app/src/main/java/ai/openclaw/app/gateway/DeviceAuthStore.kt
package ai.openclaw.app.gateway

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

data class DeviceAuthEntry(
  val token: String,
  val role: String,
  val scopes: List<String>,
  val updatedAtMs: Long,
)

@Serializable
private data class PersistedDeviceAuthEntry(
  val token: String,
  val role: String,
  val scopes: List<String> = emptyList(),
  val updatedAtMs: Long = 0L,
)

@Serializable
private data class PersistedDeviceAuthStore(
  val version: Int = 1,
  val deviceId: String,
  val tokens: Map<String, PersistedDeviceAuthEntry> = emptyMap(),
)

interface DeviceAuthTokenStore {
  fun loadEntry(
    deviceId: String,
    role: String,
  ): DeviceAuthEntry?

  fun loadToken(
    deviceId: String,
    role: String,
  ): String? = loadEntry(deviceId, role)?.token

  fun saveToken(
    deviceId: String,
    role: String,
    token: String,
    scopes: List<String> = emptyList(),
  )

  fun clearToken(
    deviceId: String,
    role: String,
  )
}

class DeviceAuthStore(
  context: Context,
) : DeviceAuthTokenStore {
  private val json = Json { ignoreUnknownKeys = true }
  private val kvStore = OpenClawSQLiteKVStore(context)

  override fun loadEntry(
    deviceId: String,
    role: String,
  ): DeviceAuthEntry? {
    val store = readStore() ?: return null
    val normalizedDevice = normalizeDeviceId(deviceId)
    if (store.deviceId != normalizedDevice) return null
    val normalizedRole = normalizeRole(role)
    val entry = store.tokens[normalizedRole] ?: return null
    val token = entry.token.trim().takeIf { it.isNotEmpty() } ?: return null
    return DeviceAuthEntry(
      token = token,
      role = normalizedRole,
      scopes = normalizeScopes(entry.scopes),
      updatedAtMs = entry.updatedAtMs,
    )
  }

  override fun saveToken(
    deviceId: String,
    role: String,
    token: String,
    scopes: List<String>,
  ) {
    val normalizedDevice = normalizeDeviceId(deviceId)
    val normalizedRole = normalizeRole(role)
    val normalizedScopes = normalizeScopes(scopes)
    val existing = readStore()
    val nextTokens =
      if (existing?.deviceId == normalizedDevice) {
        existing.tokens.toMutableMap()
      } else {
        mutableMapOf()
      }
    nextTokens[normalizedRole] =
      PersistedDeviceAuthEntry(
        token = token.trim(),
        role = normalizedRole,
        scopes = normalizedScopes,
        updatedAtMs = System.currentTimeMillis(),
      )
    writeStore(
      PersistedDeviceAuthStore(
        deviceId = normalizedDevice,
        tokens = nextTokens,
      ),
    )
  }

  override fun clearToken(
    deviceId: String,
    role: String,
  ) {
    val normalizedDevice = normalizeDeviceId(deviceId)
    val existing = readStore() ?: return
    if (existing.deviceId != normalizedDevice) return
    val normalizedRole = normalizeRole(role)
    if (!existing.tokens.containsKey(normalizedRole)) return
    val nextTokens = existing.tokens.toMutableMap()
    nextTokens.remove(normalizedRole)
    writeStore(existing.copy(tokens = nextTokens))
  }

  private fun readStore(): PersistedDeviceAuthStore? {
    val raw = kvStore.readString(DEVICE_AUTH_SCOPE, DEVICE_AUTH_KEY) ?: return null
    return runCatching { json.decodeFromString<PersistedDeviceAuthStore>(raw) }
      .getOrNull()
      ?.takeIf { it.version == 1 && it.deviceId.isNotBlank() }
  }

  private fun writeStore(store: PersistedDeviceAuthStore) {
    kvStore.writeString(
      DEVICE_AUTH_SCOPE,
      DEVICE_AUTH_KEY,
      json.encodeToString(store),
    )
  }

  private fun normalizeDeviceId(deviceId: String): String = deviceId.trim().lowercase()

  private fun normalizeRole(role: String): String = role.trim().lowercase()

  private fun normalizeScopes(scopes: List<String>): List<String> =
    scopes
      .map { it.trim() }
      .filter { it.isNotEmpty() }
      .distinct()
      .sorted()

  companion object {
    private const val DEVICE_AUTH_SCOPE = "identity.device-auth"
    private const val DEVICE_AUTH_KEY = "default"
  }
}

File: apps/shared/OpenClawKit/Sources/OpenClawKit/DeviceAuthStore.swift
import Foundation

public struct DeviceAuthEntry: Codable, Sendable {
    public let token: String
    public let role: String
    public let scopes: [String]
    public let updatedAtMs: Int

    public init(token: String, role: String, scopes: [String], updatedAtMs: Int) {
        self.token = token
        self.role = role
        self.scopes = scopes
        self.updatedAtMs = updatedAtMs
    }
}

private struct DeviceAuthStoreFile: Codable {
    var version: Int
    var deviceId: String
    var tokens: [String: DeviceAuthEntry]
}

public enum DeviceAuthStore {
    private static let stateScope = "identity.device-auth"
    private static let stateKey = "default"

    public static func loadToken(deviceId: String, role: String) -> DeviceAuthEntry? {
        guard let store = readStore(), store.deviceId == deviceId else { return nil }
        let role = self.normalizeRole(role)
        return store.tokens[role]
    }

    public static func storeToken(
        deviceId: String,
        role: String,
        token: String,
        scopes: [String] = []) -> DeviceAuthEntry
    {
        let normalizedRole = self.normalizeRole(role)
        var next = self.readStore()
        if next?.deviceId != deviceId {
            next = DeviceAuthStoreFile(version: 1, deviceId: deviceId, tokens: [:])
        }
        let entry = DeviceAuthEntry(
            token: token,
            role: normalizedRole,
            scopes: normalizeScopes(scopes),
            updatedAtMs: Int(Date().timeIntervalSince1970 * 1000))
        if next == nil {
            next = DeviceAuthStoreFile(version: 1, deviceId: deviceId, tokens: [:])
        }
        next?.tokens[normalizedRole] = entry
        if let store = next {
            self.writeStore(store)
        }
        return entry
    }

    public static func clearToken(deviceId: String, role: String) {
        guard var store = readStore(), store.deviceId == deviceId else { return }
        let normalizedRole = self.normalizeRole(role)
        guard store.tokens[normalizedRole] != nil else { return }
        store.tokens.removeValue(forKey: normalizedRole)
        self.writeStore(store)
    }

    private static func normalizeRole(_ role: String) -> String {
        role.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func normalizeScopes(_ scopes: [String]) -> [String] {
        let trimmed = scopes
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        return Array(Set(trimmed)).sorted()
    }

    private static func readStore() -> DeviceAuthStoreFile? {
        guard let data = OpenClawSQLiteKVStore.readJSONData(scope: self.stateScope, key: self.stateKey) else {
            return nil
        }
        guard let decoded = try? JSONDecoder().decode(DeviceAuthStoreFile.self, from: data) else {
            return nil
        }
        guard decoded.version == 1 else { return nil }
        return decoded
    }

    private static func writeStore(_ store: DeviceAuthStoreFile) {
        do {
            let data = try JSONEncoder().encode(store)
            try OpenClawSQLiteKVStore.writeJSONData(scope: self.stateScope, key: self.stateKey, data: data)
        } catch {
            // best-effort only
        }
    }
}
